File: pdftransl/parsing/splitter.py
"""Split Markdown into typed structural blocks.

The splitter is intentionally conservative: anything it cannot
classify stays a paragraph (translatable) and formula protection is
handled later by the masking layer, so misclassification never
corrupts math.
"""

from __future__ import annotations

import re

from pdftransl.models import Block, BlockType, TRANSLATABLE_TYPES

_HEADING_RE = re.compile(r"^#{1,6}\s")
_IMAGE_ONLY_RE = re.compile(r"^\s*!\[[^\]]*\]\([^)]*\)\s*$")
_TABLE_ROW_RE = re.compile(r"^\s*\|")
_FENCE_RE = re.compile(r"^\s*(```|~~~)")
_DISPLAY_MATH_LINE_RE = re.compile(r"^\s*\$\$")
_HTML_BLOCK_RE = re.compile(r"^\s*</?(?:div|table|tr|td|th|img|figure|p|span|br|hr)\b", re.I)
_BEGIN_ENV_RE = re.compile(r"^\s*\\begin\{([a-zA-Z*]+)\}")


def _make_block(btype: BlockType, lines: list[str], index: int) -> Block:
    return Block(
        type=btype,
        text="\n".join(lines).rstrip("\n"),
        index=index,
        translatable=btype in TRANSLATABLE_TYPES,
    )


def split_markdown(markdown: str) -> list[Block]:
    """Split a Markdown document into an ordered list of blocks."""
    lines = markdown.splitlines()
    blocks: list[Block] = []
    buf: list[str] = []
    buf_type: BlockType | None = None
    fence_marker: str | None = None
    math_open = False
    env_name: str | None = None

    def flush(btype: BlockType | None = None) -> None:
        nonlocal buf, buf_type
        if buf:
            blocks.append(_make_block(btype or buf_type or BlockType.PARAGRAPH, buf, len(blocks)))
        buf = []
        buf_type = None

    for line in lines:
        # --- multi-line states -------------------------------------
        if fence_marker is not None:
            buf.append(line)
            if line.strip().startswith(fence_marker):
                fence_marker = None
                flush(BlockType.CODE)
            continue
        if math_open:
            buf.append(line)
            # closing $$ (line containing the second delimiter)
            if "$$" in line:
                math_open = False
                flush(BlockType.MATH)
            continue
        if env_name is not None:
            buf.append(line)
            if re.search(r"\\end\{" + re.escape(env_name) + r"\}", line):
                env_name = None
                flush(BlockType.MATH)
            continue

        stripped = line.strip()

        # --- state openers ------------------------------------------
        m = _FENCE_RE.match(line)
        if m:
            flush()
            buf.append(line)
            fence_marker = m.group(1)
            continue
        if _DISPLAY_MATH_LINE_RE.match(line):
            flush()
            buf.append(line)
            # single-line $$...$$ ?
            if stripped.count("$$") >= 2:
                flush(BlockType.MATH)
            else:
                math_open = True
            continue
        m = _BEGIN_ENV_RE.match(line)
        if m:
            flush()
            buf.append(line)
            name = m.group(1)
            if re.search(r"\\end\{" + re.escape(name) + r"\}", line):
                flush(BlockType.MATH)
            else:
                env_name = name
            continue

        # --- single-line / grouped constructs ------------------------
        if not stripped:
            flush()
            continue
        if _HEADING_RE.match(line):
            flush()
            buf.append(line)
            flush(BlockType.HEADING)
            continue
        if _IMAGE_ONLY_RE.match(line):
            flush()
            buf.append(line)
            flush(BlockType.IMAGE)
            continue
        if _TABLE_ROW_RE.match(line):
            if buf_type != BlockType.TABLE:
                flush()
                buf_type = BlockType.TABLE
            buf.append(line)
            continue
        if _HTML_BLOCK_RE.match(line) and buf_type is None and not buf:
            buf.append(line)
            flush(BlockType.HTML)
            continue

        # default: paragraph text
        if buf_type == BlockType.TABLE:
            flush()
        if buf_type is None:
            buf_type = BlockType.PARAGRAPH
        buf.append(line)

    flush()
    return blocks


def assemble(blocks_text: list[str]) -> str:
    """Join block texts back into a Markdown document."""
    return "\n\n".join(t for t in blocks_text if t.strip() != "") + "\n"


_REFERENCES_HEADING_RE = re.compile(
    r"^#{1,6}\s*(?:\d+\.?\s*)?(references|bibliography|список литературы|литература|"
    r"cited works|works cited)\s*$",
    re.IGNORECASE,
)
# Headings that end the references section (appendices etc.)
_POST_REFERENCES_RE = re.compile(
    r"^#{1,6}\s*(?:[A-Z]?\d*\.?\s*)?(appendix|supplementary|acknowledg|приложение)",
    re.IGNORECASE,
)


def mark_references(blocks: list[Block]) -> int:
    """Mark the bibliography section as non-translatable.

    Reference entries (authors, titles, venues) should stay in the
    original language — translating them breaks citability. Returns
    the number of blocks marked.
    """
    marked = 0
    in_refs = False
    for block in blocks:
        if block.type == BlockType.HEADING:
            if _REFERENCES_HEADING_RE.match(block.text.strip()):
                in_refs = True
                continue  # the heading itself stays translatable
            if in_refs and _POST_REFERENCES_RE.match(block.text.strip()):
                in_refs = False
        elif in_refs and block.translatable:
            block.translatable = False
            block.meta["skipped"] = "references"
            marked += 1
    return marked

File: cpp/src/parsing/splitter.cpp
#include "parsing/splitter.h"
#include <QRegularExpression>
#include <optional>

namespace pdftransl {

namespace {

const QRegularExpression& headingRe() {
    static const QRegularExpression re(QStringLiteral("^#{1,6}\\s"));
    return re;
}
const QRegularExpression& imageOnlyRe() {
    static const QRegularExpression re(QStringLiteral("^\\s*!\\[[^\\]]*\\]\\([^)]*\\)\\s*$"));
    return re;
}
const QRegularExpression& tableRowRe() {
    static const QRegularExpression re(QStringLiteral("^\\s*\\|"));
    return re;
}
const QRegularExpression& fenceRe() {
    static const QRegularExpression re(QStringLiteral("^\\s*(```|~~~)"));
    return re;
}
const QRegularExpression& displayMathLineRe() {
    static const QRegularExpression re(QStringLiteral("^\\s*\\$\\$"));
    return re;
}
const QRegularExpression& htmlBlockRe() {
    static const QRegularExpression re(
        QStringLiteral("^\\s*</?(?:div|table|tr|td|th|img|figure|p|span|br|hr)\\b"),
        QRegularExpression::CaseInsensitiveOption);
    return re;
}
const QRegularExpression& beginEnvRe() {
    static const QRegularExpression re(QStringLiteral("^\\s*\\\\begin\\{([a-zA-Z*]+)\\}"));
    return re;
}
const QRegularExpression& referencesHeadingRe() {
    static const QRegularExpression re(
        QStringLiteral("^#{1,6}\\s*(?:\\d+\\.?\\s*)?(references|bibliography|"
                        "список литературы|литература|cited works|works cited)\\s*$"),
        QRegularExpression::CaseInsensitiveOption);
    return re;
}

QString normalizeNewlines(const QString& text) {
    QString out = text;
    out.replace(QStringLiteral("\r\n"), QStringLiteral("\n"));
    out.replace(QChar('\r'), QChar('\n'));
    return out;
}

} // namespace

bool isTranslatableType(BlockType type) {
    return type == BlockType::Text || type == BlockType::Heading || type == BlockType::Table;
}

std::vector<Block> splitMarkdown(const QString& markdown) {
    std::vector<Block> blocks;
    const QStringList lines = normalizeNewlines(markdown).split('\n');

    QStringList buf;
    std::optional<BlockType> bufType;
    QString fenceMarker;   // непусто, пока мы внутри ``` / ~~~ ограды
    bool mathOpen = false; // внутри многострочного $$ ... $$
    QString envName;       // непусто, пока мы внутри \begin{name} ... \end{name}

    auto flush = [&](std::optional<BlockType> forced = std::nullopt) {
        if (!buf.isEmpty()) {
            Block block;
            block.type = forced ? *forced : bufType.value_or(BlockType::Text);
            block.text = buf.join('\n');
            block.page = 0;
            blocks.push_back(std::move(block));
        }
        buf.clear();
        bufType.reset();
    };

    for (const QString& line : lines) {
        // --- многострочные состояния -------------------------------
        if (!fenceMarker.isEmpty()) {
            buf << line;
            if (line.trimmed().startsWith(fenceMarker)) {
                fenceMarker.clear();
                flush(BlockType::Code);
            }
            continue;
        }
        if (mathOpen) {
            buf << line;
            if (line.contains(QStringLiteral("$$"))) {
                mathOpen = false;
                flush(BlockType::Math);
            }
            continue;
        }
        if (!envName.isEmpty()) {
            buf << line;
            QRegularExpression endRe(QStringLiteral("\\\\end\\{") + QRegularExpression::escape(envName) +
                                      QStringLiteral("\\}"));
            if (endRe.match(line).hasMatch()) {
                envName.clear();
                flush(BlockType::Math);
            }
            continue;
        }

        const QString stripped = line.trimmed();

        // --- открытие многострочных состояний -----------------------
        QRegularExpressionMatch fenceMatch = fenceRe().match(line);
        if (fenceMatch.hasMatch()) {
            flush();
            buf << line;
            fenceMarker = fenceMatch.captured(1);
            continue;
        }
        if (displayMathLineRe().match(line).hasMatch()) {
            flush();
            buf << line;
            if (stripped.count(QStringLiteral("$$")) >= 2) {
                flush(BlockType::Math);
            } else {
                mathOpen = true;
            }
            continue;
        }
        QRegularExpressionMatch envMatch = beginEnvRe().match(line);
        if (envMatch.hasMatch()) {
            flush();
            buf << line;
            const QString name = envMatch.captured(1);
            QRegularExpression endRe(QStringLiteral("\\\\end\\{") + QRegularExpression::escape(name) +
                                      QStringLiteral("\\}"));
            if (endRe.match(line).hasMatch()) {
                flush(BlockType::Math);
            } else {
                envName = name;
            }
            continue;
        }

        // --- однострочные / группируемые конструкции -----------------
        if (stripped.isEmpty()) {
            flush();
            continue;
        }
        if (headingRe().match(line).hasMatch()) {
            flush();
            buf << line;
            flush(BlockType::Heading);
            continue;
        }
        if (imageOnlyRe().match(line).hasMatch()) {
            flush();
            buf << line;
            flush(BlockType::Image);
            continue;
        }
        if (tableRowRe().match(line).hasMatch()) {
            if (bufType != BlockType::Table) flush();
            bufType = BlockType::Table;
            buf << line;
            continue;
        }
        if (buf.isEmpty() && !bufType.has_value() && htmlBlockRe().match(line).hasMatch()) {
            buf << line;
            flush(BlockType::Html);
            continue;
        }

        // умолчание: текст абзаца
        if (bufType == BlockType::Table) flush();
        if (!bufType.has_value()) bufType = BlockType::Text;
        buf << line;
    }
    flush();
    return blocks;
}

std::vector<Block> markReferences(const std::vector<Block>& blocksIn, int* outMarked) {
    std::vector<Block> blocks = blocksIn;
    int marked = 0;
    bool inRefs = false;
    for (auto& block : blocks) {
        if (block.type == BlockType::Heading) {
            inRefs = referencesHeadingRe().match(block.text.trimmed()).hasMatch();
        } else if (inRefs && isTranslatableType(block.type)) {
            block.type = BlockType::References;
            ++marked;
        }
    }
    if (outMarked) *outMarked = marked;
    return blocks;
}

std::vector<Segment> toSegments(const std::vector<Block>& blocks) {
    std::vector<Segment> segments;
    segments.reserve(blocks.size());
    for (const auto& block : blocks) {
        Segment segment;
        segment.id = newId("seg_");
        segment.kind = block.type;
        segment.sourceText = block.text;
        segment.passthrough = !isTranslatableType(block.type) || block.text.trimmed().isEmpty();
        segments.push_back(std::move(segment));
    }
    return segments;
}

QString assemble(const std::vector<QString>& texts) {
    QStringList nonEmpty;
    nonEmpty.reserve(static_cast<int>(texts.size()));
    for (const auto& text : texts) {
        if (!text.trimmed().isEmpty()) nonEmpty << text;
    }
    return nonEmpty.join(QStringLiteral("\n\n")) + QStringLiteral("\n");
}

QString assemble(const std::vector<Segment>& segments) {
    std::vector<QString> texts;
    texts.reserve(segments.size());
    for (const auto& segment : segments) texts.push_back(segment.finalText());
    return assemble(texts);
}

} // namespace pdftransl
